planner/types.go

```diff
@@ -18,11 +18,41 @@ type PlanNode interface {
 
 	// Helper methods for manipulating a plan
 	// These methods are used during planning
-	AddSuccessor(PlanNode)
-	AddPredecessor(PlanNode)
+	AddSuccessors(...PlanNode)
+	AddPredecessors(...PlanNode)
 	RemovePredecessor(PlanNode)
 }
 
+// QueryPlan holds the roots of the query plan DAG.
+// Roots correspond to nodes that produce final results.
+type QueryPlan struct {
+	roots map[PlanNode]struct{}
+}
+
+// NewQueryPlan instantiates a new query plan
+func NewQueryPlan(nodes []PlanNode) *QueryPlan {
```

Comment:
This argument would be better off called `roots` I think, lest callers think that they need to pass a slice of every node in the plan.

Reply:
True, I'll fix.